File: Graph/30NumberOfIslands/number_of_islands.py
class Solution:
    def while_it_has_neighbors(self, grid, row, col, ROW, COL, is_visited):
        if (row < 0 or col < 0 or row == ROW or col == COL):
            return
        if (grid[row][col] == '0' or is_visited[row][col]):
            is_visited[row][col] = True
            return
        is_visited[row][col] = True
        self.while_it_has_neighbors(grid, row - 1, col, ROW, COL, is_visited)
        self.while_it_has_neighbors(grid, row, col + 1, ROW, COL, is_visited)
        self.while_it_has_neighbors(grid, row + 1, col, ROW, COL, is_visited)
        self.while_it_has_neighbors(grid, row, col - 1, ROW, COL, is_visited)
        return

    def numIslands(self, grid: List[List[str]]) -> int:
        ROW, COL = len(grid), len(grid[0])
        res = 0
        is_visited = [[False for _ in range(COL)] for _ in range(ROW)]
        for row in range(ROW):
            for col in range(COL):
                if (grid[row][col] == '0' or is_visited[row][col]):
                    is_visited[row][col] = True
                    continue
                else:
                    self.while_it_has_neighbors(grid, row, col, ROW, COL, is_visited)
                    res += 1
        return res

File: Graph/30NumberOfIslands/NumberOfIslands.java
class Solution {
    public void itHasNeighbors(char[][] grid, int row, int col, int ROW, int COL, boolean[][] isVisited) {
        if (row < 0 || col < 0 || col == COL || row == ROW) {
            return;
        }
        if (grid[row][col] == '0' || isVisited[row][col]) {
            isVisited[row][col] = true;
            return;
        }
        isVisited[row][col] = true;
        itHasNeighbors(grid, row - 1, col, ROW, COL, isVisited);
        itHasNeighbors(grid, row, col + 1, ROW, COL, isVisited);
        itHasNeighbors(grid, row + 1, col, ROW, COL, isVisited);
        itHasNeighbors(grid, row, col - 1, ROW, COL, isVisited);
        return;
    }

    public int numIslands(char[][] grid) {
        boolean[][] isVisited = new boolean[grid.length][grid[0].length];
        int ROW = grid.length, COL = grid[0].length, res = 0;
        for (int row = 0; row < ROW; row++) {
            for (int col = 0; col < COL; col++) {
                if (grid[row][col] == '0' || isVisited[row][col]) {
                    isVisited[row][col] = true;
                    continue;
                } else {
                    itHasNeighbors(grid, row, col, ROW, COL, isVisited);
                    res += 1;
                }
            }
        }
        return res;
    }
}
